### DISCUSSION TITLE
Examples namings.

### DISCUSSION THREAD

**Discussion by uwejan** [libp2p/rust-libp2p#3609] [Q&A] [answered]
Hello there,

Just cloned the repo, and am trying to run the examples.
```
cargo build --example 
error: "--example" takes one argument.
Available examples:
    client
    listen_ping
    metrics
    relay
    server
```
Is it supposed to reflect on the directory `examples` on the main repo?

Now following tutorial from docs, does not use the same naming for examples.
https://docs.rs/libp2p/latest/libp2p/tutorials/hole_punching/index.html
```
./client --secret-key-seed 1 --mode listen --relay-address /ip4/$RELAY_SERVER_IP/tcp/4001/p2p/12D3KooWDpJ7As7BWAwRMfu1VU2WCqNjvq387JEYKDBj4kx6nXTN


error: unexpected argument '--secret-key-seed' found

  note: argument '--server-peer-id' exists

Usage: client <--listen-port <LISTEN_PORT>|--server-address <SERVER_ADDRESS>|--server-peer-id <SERVER_PEER_ID>>

For more information, try '--help'.
````

protocoles directory, has also examples. 

This somehow causing confusion where example are pulled from.

**Comment by thomaseizinger** [comment, marked as answer]
Thanks for reporting!

We have recently moved all examples to make things more consistent, see https://github.com/libp2p/rust-libp2p/issues/3111.

All our examples are dedicated crates to also highlight which dependencies are needed. You should be able to `cd` into any of these directories and do `cargo run`: https://github.com/libp2p/rust-libp2p/tree/master/examples

The rustdoc on top of the `main` function on all of them also some some more documentation.

> Now following tutorial from docs, does not use the same naming for examples. [docs.rs/libp2p/latest/libp2p/tutorials/hole_punching/index.html](https://docs.rs/libp2p/latest/libp2p/tutorials/hole_punching/index.html)

This might still be outdated from the move :(
PRs to fix it are appreciated!

**Reply by uwejan** [reply]
Got it. Thank you.

**Reply by uwejan** [reply]
Hello @thomaseizinger , Back with some questions;
Following up with this example; https://github.com/libp2p/rust-libp2p/blob/master/examples/rendezvous/src/main.rs#L21
I could not make it work, I am suspecting the port need to be static here, https://github.com/libp2p/rust-libp2p/blob/master/examples/rendezvous/src/main.rs#L75, then changed accordingley for other files in `bin` directory.

I been reading the specs here https://github.com/libp2p/specs/tree/master/rendezvous#interaction
Will the server be acting as relay/fan-out OR is it supposed to upgrade the connection to a direct connection between A and B, after exchanging the best rout to connect? Perhabs am missing out a concept here or there, I appreciate your directions.

Does the same concept as I explained before apply for https://docs.rs/libp2p/latest/libp2p/tutorials/hole_punching/index.html as well? 


**I could sum my use case as follow: 
Peers( A, B); could be on local LAN, or on WAN behind NAT. (OR more complex use case on virtual enviroment but lets ignore that for now).
The goal is for A,B have a direct websocekt connection eventually.** 
The following implementation https://github.com/optman/rndz , will upgrade the connection to a direct connection eventually.
That leaves me with mDNS to handle Local LAN.

But the question remains here, How can `libp2p2` be of usage for me, which I am trying to figure out.

Thank you.

**Reply by thomaseizinger** [reply]
> Following up with this example; [`master`/examples/rendezvous/src/main.rs#L21](https://github.com/libp2p/rust-libp2p/blob/master/examples/rendezvous/src/main.rs#L21)
> I could not make it work, I am suspecting the port need to be static here, [`master`/examples/rendezvous/src/main.rs#L75](https://github.com/libp2p/rust-libp2p/blob/master/examples/rendezvous/src/main.rs#L75), then changed accordingley for other files in `bin` directory.

Yes you are correct. The port in `main.rs` needs to be changed to 62649.

> I been reading the specs here [libp2p/specs@`master`/rendezvous#interaction](https://github.com/libp2p/specs/tree/master/rendezvous#interaction) Will the server be acting as relay/fan-out OR is it supposed to upgrade the connection to a direct connection between A and B, after exchanging the best rout to connect? Perhabs am missing out a concept here or there, I appreciate your directions.

Rendezvous does not perform any connection upgrades. It only acts as a small database. A service discovery if you want. One service can register themselves for a certain namespace and others can discover it. Once you have discovered a node, you can then connect to it.

It _could_ be that a node registered themselves with a _relay_ address (if they cannot listen on a publicly reachable port) in which case you could then connect to it via the relay and potentially holepunch them via the _dcutr_ protocol.

> Does the same concept as I explained before apply for [docs.rs/libp2p/latest/libp2p/tutorials/hole_punching/index.html](https://docs.rs/libp2p/latest/libp2p/tutorials/hole_punching/index.html) as well?
> 
> **I could sum my use case as follow: Peers( A, B); could be on local LAN, or on WAN behind NAT. (OR more complex use case on virtual enviroment but lets ignore that for now). The goal is for A,B have a direct websocekt connection eventually.** The following implementation [optman/rndz](https://github.com/optman/rndz) , will upgrade the connection to a direct connection eventually. That leaves me with mDNS to handle Local LAN.
> 
> But the question remains here, How can `libp2p2` be of usage for me, which I am trying to figure out.

If I understand you correctly, something like the following might help:

- Node A wants to offer a service to other nodes
- It starts listening on an address
- Use the AutoNAT protocol to figure out its NAT status
- If behind NAT:
	- Chose a relay from a predefined list and listen on that
	- Publish the relayed address to a predefined rendezvous node
If not behind NAT:
	- Publish the regular listen address to a predefined rendezvous node

You could also add mDNS to the mix here. Basically, you can just dial any node that you discover via mDNS and wait for the identify handshake to complete to see which protocols they support.

**Reply by uwejan** [reply]
Thank you Thomas, this has been informative.
Going to reflect on it for a bit, perhabs will come back some times later with some questions.
Thank you.
